[nomerge] Optimise equals checking for IndexedSeq - 2.12 only

mkeskells opened the pull request:
Optimise implementation of equals for IndexedSeq

without this `equals` creates two iterators and walks the content

mimics https://github.com/scala/scala/pull/6833 but without binary changes

mkeskells committed:
Optimise equals checking for Vector

mkeskells commented:
relates to https://github.com/rorygraves/scalac_perf/issues/75

retronym commented:
I would target this at `IndexedSeqOptimized`, or maybe just `Vector`, in 2.12.x.

mkeskells commented:
@retronym which would you prefer. The production issue we observed is Vector specific

retronym commented:
To elaborate, the code as written has two problems as I see it:

  - It doesn't call `canEqual`, so a custom collection extending `IndexedSeq` that opts out of seq equality might incorrectly equate to a general seq.
  - It assumes `IndexedSeq.apply` is `O(1)` (or effectively so), whereas in 2.12.x that property comes further down in `IndexedSeqOptimized`.

Let's push this down to `Vector.equals`.

mkeskells commented:
@retronym I have moved the comparison to be based on `sameElements`
I have moved the checks to be Vector specific, but cant mve the methods to Vector as this would break binary compat wouldnt it. I presume this is what you wanted

adriaanm commented:
ping @retronym -- is this good to go?

mkeskells commented:
@adriaanm this was in response to production issue. There is a more wide-ranging change already merged into 2.13 - https://github.com/scala/scala/pull/6833

retronym commented:
Sorry I dropped the ball on this one, it's probably too late to include in 2.12.7.
